src/Mafia/Home.hs

```diff
@@ -51,30 +52,28 @@ installBinary package deps = do
   tmp <- ensureMafiaDir "tmp"
   let nv = renderPackageId package
   bin <- ensureMafiaDir "bin"
-  let path = bin </> nv
+  let prefix = bin </> nv
+      path = prefix </> "bin"
+      lock = T.unpack (prefix </> "lock")
+      marker = prefix </> "installed"
+      installArgs p = [ "install", renderPackageId p, "--prefix=" <> prefix ]
   -- This is to support old versions of mafia that had installed a single executable
   -- We can remove this at some point
   -- NOTE: This makes no attempt at being threadsafe
-  liftIO (doesFileExist path) >>= \b -> when b $ do
-    renameFile path (path <> ".tmp")
-    createDirectoryIfMissing True path
-    renameFile (path <> ".tmp") (path </> (unPackageName . pkgName) package)
-  liftIO (doesDirectoryExist path) >>= \case
-    True ->
-      pure path
-    False -> do
+  whenM (liftIO (doesFileExist prefix)) (removeFile prefix)
+  bracketEitherT' (liftIO (lockFile lock Exclusive)) (liftIO . unlockFile) $ \_ ->
+    unlessM (liftIO (doesFileExist marker)) $
```

Comment:
I think it's in the right spot
- a rival build is running and fails, we wait for the lock, see no marker file, start a new build
- rival build succeeds, we get the lock, see marker file, no need to build